### DISCUSSION TITLE
Behaviour of ScaledJob `minReplicaCount` field

### DISCUSSION THREAD

**Discussion by LewisJackson1** [kedacore/keda#4885] [General]
We've had a bit of discussion in the comments of this issue https://github.com/kedacore/keda/issues/4554#issuecomment-1680205379 concerning the behaviour of the `minReplicaCount` of the ScaledJob object. I was asked to open a discussion here to get more opinions on the matter.

Just to recap, the behaviour currently is defined as:

>  ```minReplicaCount: 10 # Optional. Default: 0```
> The min number of jobs that is created by default. This can be useful to avoid bootstrapping time of new jobs. If minReplicaCount is greater than maxReplicaCount, minReplicaCount will be set to maxReplicaCount.
> New messages may create new jobs - within the limits imposed by maxReplicaCount - in order to reach the state where minReplicaCount jobs are always running. For example, if one sets minReplicaCount to 2 then there will be 2 jobs running permanently. Using a targetValue of 1, if 3 new messages are sent, 2 of those messages will be processed on the already running jobs but another 3 jobs will be created in order to fulfill the desired state dictated by the minReplicaCount parameter that is set to 2.

In my use case, we are considering migrating a ScaledObject to a ScaledJob. In doing so I expected the `minReplicaCount` of ScaledJob to match the behaviour of a ScaledObject. So to use the example above:

> For example, if one sets minReplicaCount to 2 then there will be 2 jobs running permanently. Using a targetValue of 1, if 3 new messages are sent, 2 of those messages will be processed on the already running jobs. 1 more job will be created to process the final message.

The consequences of scaling out too many Jobs in our use case is that we'd incur minimum charges for GPU Nodes. We would like to have two Jobs kept warm without the overprovisioning behaviour that is currently intended. I can understand that this feature may be useful for some users, but I'm not sure that it's common enough to be the default.

The discussion in the issue above is worth reading and all opinions are welcome! I can think of a couple of ways to work around this behaviour to fit our use case, but it'd have been great to have it working out of the box.

**Comment by JorTurFer** [comment]
@tomkerkhove @zroubalik , please share your thoughts

**Reply by eugen-nw** [reply]
@LewisJackson1 how long does your processing typically take?  How large are the bursts that you need to scale out to in order to accommodate?  Also what drives you to move to using ScaledJobs?  Perhaps you do not need that in the first place.

In my world we can easily get > 40 Messages to process in less than a minute.  Most of them compute in < 1 minute but some of them can take hours.  Because of the long running ones I have to use ScaledJob so I am guaranteed that the processing will complete in the Pod it lands on.  I run with minReplicaCount set to 12. and max. set to 80.  The moment one Message gets into the Queue, I like the fact that the ScaledJob starts up a new Pod so I will be better prepared to handle a potential burst.  Let’s stay with this one Message example for a moment.  ScaledJob guarantees that I will have 12 Pods running, which is good.  I pay extra for the single Pod that is processing the Message.  The moment the processing is over, that Pod terminates and we no longer pay for it.  Should 20 Messages come the next minute, I have 11 Pods ready to process them, the newly launched one almost ready and ScaledJob will provide me 20 brand new ones that will accommodate the remaining 8 Messages, plus have a spare of 12 Pods to accommodate the next burst.

That’s how scale-out works, its purpose is to accommodate burst in demand in a reasonable manner.  Whenever you scale out two things will happen:  1. extra costs  2. possible delays until scale-out becomes functional.  ScaledJob reduces number 2 because it starts up new Pods fast.  ScaledObject is slower with the scale-out aspect and can easily scale out to more Pods than necessary.

**Reply by eugen-nw** [reply]
On the other hand if you want to minimize costs, use ScaledJob and have minReplicaCount set to 0.  Whenever you get N Messages, N Pods will start up to process them and will shut down when the processing is over.

**Reply by LewisJackson1** [reply]
I appreciate the in-depth example but I feel that we're focusing on how the current behaviour could work well for some users. I don't think it's fair to say that this behaviour is how scale-out works - it's just how it's been defined by whoever wrote the PR. My point is actually that I think this is an abnormal behaviour for scaling out a set of Jobs. I don't think if we were discussing how to design this feature from the start that we'd all agree that this behaviour is the obvious default.

Some people may take a _"better safe than sorry"_ approach here and want to scale out warm Pods in case more items hit the queue. In our use case we've got two GPU Nodes reserved for a year so we may as well keep warm Pods running on those, but we're almost always going to be wasting money scaling out to `queuelength + 2` Pods just in case another Job comes in.

**Reply by eugen-nw** [reply]
If 2 is your desired count of warm Pods, you may want to use a ScaledObject with minReplicaCount set to 2 and triggers set to metricType: Value with a high messageCount: ‘5’ for example.  Or even higher than that?  ScaledObject will think for a couple of minutes before it scales out and scales out slowly, in progressively larger increments.

I still don’t understand if you actually need to have the ability to scale out to accommodate higher loads or your point is to have only 2 Pods running.  If it is the latter, then you do not need to use any scale-out at all.  Just let the 2 Pods consume whatever load comes in at the rate that they can accomplish.

I am on the side of the spectrum where I need to have warm Pods ready to handle whatever random loads that get thrown at our solution.  That is the whole point of scaling out.

**Reply by LewisJackson1** [reply]
> If 2 is your desired count of warm Pods, you may want to use a ScaledObject with minReplicaCount set to 2 and triggers set to metricType: Value with a high messageCount: ‘5’ for example. Or even higher than that? ScaledObject will think for a couple of minutes before it scales out and scales out slowly, in progressively larger increments.

We were using ScaledObject before but the behaviour described [here](https://keda.sh/docs/2.11/concepts/scaling-deployments/#long-running-executions) made it quite wasteful:

> One important consideration to make is how this pattern can work with long-running executions. Imagine a deployment triggers on a RabbitMQ queue message. Each message takes 3 hours to process. It’s possible that if many queue messages arrive, KEDA will help drive scaling out to many replicas - let’s say 4. Now the HPA makes a decision to scale down from 4 replicas to 2. There is no way to control which of the 2 replicas get terminated to scale down. That means the HPA may attempt to terminate a replica that is 2.9 hours into processing a 3 hour queue message.

It'd mean that if we'd scaled out to 4 Pods and one Pod finished a job, there'd be a 25% chance that that Pod would be the one to be terminated.

> I still don’t understand if you actually need to have the ability to scale out to accommodate higher loads or your point is to have only 2 Pods running. If it is the latter, then you do not need to use any scale-out at all. Just let the 2 Pods consume whatever load comes in at the rate that they can accomplish.

We need to process videos from our users which could take less than a minute, 30 minutes, or a few hours. I don't think it matters too much about our particular use case here as I just wanted to discuss the current behaviour of the scaler.

> I am on the side of the spectrum where I need to have warm Pods ready to handle whatever random loads that get thrown at our solution. That is the whole point of scaling out.

Again, I disagree that the point of scaling out is to have Pods sitting without work to do. I understand that there's a desire for this but I don't think it is accurate to say that this is the whole point of scaling. I think that a large percentage of users would just expect Pods to spin up as necessary.

The behaviour of having a buffer of two Pods sitting warm at all times can easily be replicated by deploying a Deployment with 2 replicas to process Jobs, or a separate ScaledJob with a min/max of 2/2. I don't think there's a simple way to get back to the behaviour that I would describe as normal.

**Reply by deanpatel2** [reply]
@LewisJackson1 

The `default` scalingStrategy uses this formula:

`maxScale - runningJobCount`


Doesn't this mean KEDA takes into account the running jobs and wouldn't cause over-provisioning for your use case? So with 5 messages (maxScale = 5) and 2 warm Jobs (runningJobCount = 2), KEDA would create 3 new Jobs, not 5.

I have a similar use case to you, but my interpretation of the default scalingStrategy is that it does satisfy our case. Perhaps this was not the default 2 years ago.

**Reply by LewisJackson1** [reply]
I've not tried to use this feature since my job at the time of opening this ticket. But the behaviour at the time was as described.

The example that you've given doesn't tackle the issues described above where the minimum replica count always wasted resources. You'd need to pick numbers where maximum replica count is large enough to have Pods doing nothing.

**Reply by deanpatel2** [reply]
Does the example I listed above not result in every pod performing work? The reason you can get into the case where Pods do nothing is if you have 2 warm pods, and when 5 messages come in, 5 brand new pods are created. Then you have a total of 7 pods for 5 messages. My example describes the case where you get the desired 5 Pods?

**Reply by LewisJackson1** [reply]
Sorry, not been keeping up with this thread and I have to keep rereading it to remember what the problem is.

7 is more than 5, you are wasting 2 Pods. I believe there should only be 5 Pods. To get a buffer of 2 Pods at all times is possible by adding another Deployment that always runs 2 Pods without scaling.

Again, my argument here is that you're trying to work backwards from what the logic is now. Imagine if someone told you they wanted this feature - I am arguing that you would not come up with this logic. We're all trying to justify one person's weird decision here.

**Comment by jasonrberk** [comment]
I stumbled onto this thread while trying to wrap my head around using keda and I feel like I'm now just more confused :-)

If you're using a scaled job, the intent is for the job to pull a message, process it and terminate.  If you start up the so-called warm instances, won't they just hit the time out more often than not before a message lands on the queue and terminate?

**Reply by ghs** [reply]
Exactly same here. Have been scratching my head for some time now and after reading here, from one side it makes sense but from another its totally confusing and looks wrong...

In our case, we have a queue where services send messages to. From the other, we use ScaledJob with the following settings:

```
      pollingInterval            = 60
      maxReplicaCount            = 1 # this is 1 because otherwise goes crazy
      successfulJobsHistoryLimit = 1
      failedJobsHistoryLimit     = 5
      jobTargetRef = {
        completions           = 30
        parallelism           = 30
        activeDeadlineSeconds = 60
```

If `maxReplicaCount` is set to 1, it will start ONE job with 30 pods in parallel. We had an accident where we had maxReplicaCount not set and basically burned our cluster... xD
The problem with this comes that if there is ONE message in the queue, we have 29 pods doing absolutely nothing and after 60 seconds, they will be marked as failed, thus, 1 success + 29 failed = job failed.

From the documentation, the scaling strategy is mainly (or only) focused on the maxReplicaCount + jobs, but not the actual messages available at the moment of pulling.

Shouldn't there be an option to choose if one really one to have warm pods or not? I am with the same thinking as @LewisJackson1 
and in addition to @jasonrberk regarding the termination problem. We've noticed that if a pod starts its processing late, because a message could arrive at some point, and hits the deadline time, it will be basically terminated in the middle of the work creating inconsistencies in data and processes related to it...


@JorTurFer @tomkerkhove @zroubalik @eugen-nw

**Reply by JorTurFer** [reply]
> If maxReplicaCount is set to 1, it will start ONE job with 30 pods in parallel. We had an accident where we had maxReplicaCount not set and basically burned our cluster... xD

But this is perfectly normal because you have set `parallelism = 30` and `completions  = 30`, This isn't related with KEDA, if you create a job (regular k8s job) with that config you will see the same 30 pods at once.

> The problem with this comes that if there is ONE message in the queue, we have 29 pods doing absolutely nothing and after 60 seconds, they will be marked as failed, thus, 1 success + 29 failed = job failed.

This is because the job spec creates 30 pods at once.

I'd suggest using `parallelism: 1`  and `completions  : 1`  to spin up a pod per message in the queue (assuming that you have set 1:1 in your trigger threshold)

**Reply by ghs** [reply]
@JorTurFer the main point in here as @LewisJackson1 commented among the others, is that if we use ScaledObject maxReplicaCount behaves differently compared to if we've used ScaledJob.

In addition, the whole spec for ScaledJob differs from the ScaledObject with the fact that in ScaledObject, we only specify the target and the target is where we say what and how, meanwhile for ScaledJob, we have the jobTargetRef.template plus jobTargetRef. completions and parallelism. So when one doesn't pay attention and expect things to be one way, clusters start to burn when supposedly correct values are set.

Personally for me, the ScaledJob should be just like the Object:

```
scaleTargetRef:
    apiVersion:    {api-version-of-target-resource}         # Optional. Default: apps/v1
    name:          {name-of-target-resource}                # Mandatory. Must be in the same namespace as the ScaledObject
minReplicaCount: 10                         # Optional. Default: 0
maxReplicaCount: 100
```
and then leave the job itself to specify parallelism and completions.

**Reply by LewisJackson1** [reply]
And I'd like to again highlight that the behaviour that's defined by Keda can really easily be replicated with a tiny bit of extra effort:

> The behaviour of having a buffer of two Pods sitting warm at all times can easily be replicated by deploying a Deployment with 2 replicas to process Jobs, or a separate ScaledJob with a min/max of 2/2. I don't think there's a simple way to get back to the behaviour that I would describe as normal.

**Reply by JorTurFer** [reply]
Behaviors are different, but that's intended and it's because they are for different goals. 

> Personally for me, the ScaledJob should be just like the Object:

This can't be done because ScaledObject requires items that implement /scale and batch/v1/job doesn't implement it. This would require custom logic to handle this difference and opens the door to implement custom logic for other CRDs without /scale support, adding more and more custom logic hard to maintain. The current ScaledObject relies on scale client and will work for any CRD that implements it.

About the usage of minReplica and its different behavior, you're right but it's because Job are different by definition that other regular workloads (IMHO at least)

**Reply by ghs** [reply]
> This can't be done because ScaledObject requires items that implement /scale and batch/v1/job doesn't implement it.

I meant about the syntax. One can create a job, and target it just like targeting a deployment.

**Reply by LewisJackson1** [reply]
> About the usage of minReplica and its different behavior, you're right but it's because Job are different by definition that other regular workloads (IMHO at least)

I don't want to be pushy, but as I've mentioned in this discussion a few times: anyone can add a buffer of any amount of replicas very easily with a Deployment. No one can choose to opt out of this logic if they wish to.

**Reply by JorTurFer** [reply]
I could argue that it's also easy having a deployment with 2 static replicas and use ScaledJob with minReplicaCount:0 to increase from there. Personally, I think that current difference makes sense considering that they are different APIs, if they were the same API with different behaviors, it'd be wrong for me too. 

> I meant about the syntax. One can create a job, and target it just like targeting a deployment.

But the problem is that it's just that the close thing, the syntax, in terms of scaling, they are totally different. I agree that we could support using an already existing job as template instead of requiring the job spec there and, actually, it could solve other problems inherited from the CRD limitations in regard of openapi features, but it's another discussion. If you think that this part can be useful, I suggest opening and issue requesting it

**Reply by LewisJackson1** [reply]
> I could argue that it's also easy having a deployment with 2 static replicas and use ScaledJob with minReplicaCount:0 to increase from there

This isn't true, this would replicate the current behaviour, not my desired behaviour. My desired behaviour is that if I have two minimum replicas, we don't add a new replica until both of those are busy.

If we're agreed that it's possible to opt in to the current behaviour, why not make this toggleable or use my desired behaviour with docs describing how to add a Deployment?

**Reply by JorTurFer** [reply]
The current behavior is the actual behavior atm and changing it is a breaking change, we can add a new parameter to manage the min replica behavior to support both (and I think it's a good addition tbh), but the current behavior has to be the default one.
We could add something like `minReplicaBehavior` or so with some options like: `useAll` and `reservedCapacity` (names are just a guidance) where the current behavior is the default one.

Can it fit your necessities?

**Reply by LewisJackson1** [reply]
I'm not using this feature anymore, I've just been hoping to improve the offering in this discussion! But yeah, that sounds like a good compromise to me.
